transports/quic/src/transport.rs

```diff
@@ -0,0 +1,703 @@
+use crate::{endpoint::EndpointChannel, muxer::QuicMuxer, upgrade::Connecting};
+use crate::{Config, ConnectionError};
+
+use futures::channel::{mpsc, oneshot};
+use futures::future::BoxFuture;
+use futures::ready;
+use futures::stream::StreamExt;
+use futures::{prelude::*, stream::SelectAll};
+
+use if_watch::{IfEvent, IfWatcher};
+
+use libp2p_core::{
+    multiaddr::{Multiaddr, Protocol},
+    transport::{ListenerId, TransportError as CoreTransportError, TransportEvent},
+    PeerId, Transport,
+};
+use rand::prelude::SliceRandom;
+use rand::thread_rng;
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, VecDeque};
+use std::marker::PhantomData;
+use std::net::IpAddr;
+use std::task::Waker;
+use std::{
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+#[derive(Debug)]
+pub struct GenTransport<P> {
+    config: Config,
+    listeners: SelectAll<Listener>,
+    /// Dialer for each socket family if no matching listener exists.
+    dialer: HashMap<SocketFamily, Dialer>,
+
+    _marker: PhantomData<P>,
+}
+
+impl<P> GenTransport<P> {
+    pub fn new(config: Config) -> Self {
+        Self {
+            listeners: SelectAll::new(),
+            config,
+            dialer: HashMap::new(),
+            _marker: Default::default(),
+        }
+    }
+}
+
+/// Error that can happen on the transport.
+#[derive(Debug, thiserror::Error)]
+pub enum TransportError {
+    /// Error while trying to reach a remote.
+    #[error(transparent)]
+    Reach(#[from] quinn_proto::ConnectError),
+    /// Error after the remote has been reached.
+    #[error(transparent)]
+    Established(#[from] ConnectionError),
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+
+    /// The task driving the endpoint has crashed.
+    #[error("Endpoint driver crashed")]
+    EndpointDriverCrashed,
+}
+
+impl<P: Provider> Transport for GenTransport<P> {
+    type Output = (PeerId, QuicMuxer);
+    type Error = TransportError;
+    type ListenerUpgrade = Connecting;
+    type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn listen_on(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<ListenerId, CoreTransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or(CoreTransportError::MultiaddrNotSupported(addr))?;
+        let listener_id = ListenerId::new();
+        let listener = Listener::new::<P>(listener_id, socket_addr, self.config.clone())?;
+        self.listeners.push(listener);
+
+        // Remove dialer endpoint so that the endpoint is dropped once the last
+        // connection that uses it is closed.
+        // New outbound connections will use a bidirectional (listener) endpoint.
+        self.dialer.remove(&socket_addr.ip().into());
+
+        Ok(listener_id)
+    }
+
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        if let Some(listener) = self.listeners.iter_mut().find(|l| l.listener_id == id) {
```

Comment:
Would it make sense to remove the listener? I'm not sure it's possible with `SelectAll`, but it's a bit confusing because we keep a closed listener forever (?).

Edit: nvm. I see from the test below that SelectAlll seems to prune completed streams (since that's what `FuturesUnordered` does. Maybe leave a comment here explaining this?

Reply:
Comment added in 3fa3e9ef635b991a14d49898fd72570eea612594.